src/components/TodoList.tsx

```diff
@@ -0,0 +1,270 @@
+  ) => {
+    event.preventDefault();
+    if (!todoTitle.trim()) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    setTempTodo({
+      id: 0,
+      title: todoTitle.trim(),
+      userId,
+      completed: false,
+    });
+
+    setIsInputDisabled(true);
+
+    postTodo({
+      title: todoTitle.trim(),
+      userId,
+      completed: false,
+    })
+      .then(newTodo => {
+        setTodos(currentTodos => [...currentTodos, newTodo]);
+        setTodoTitle('');
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+      })
+      .finally(() => {
+        setIsInputDisabled(false);
+        setTempTodo(null);
+      });
+  };
+
+  const handleDeleteTodo = (id: number) => {
+    setProcessingTodoIds(currentTodoIds => [...currentTodoIds, id]);
+
+    return deleteTodo(id)
+      .then(() => {
+        setTodos(currentTodos => currentTodos
+          .filter(currentTodo => currentTodo.id !== id));
+        setFocusToHeaderInput(currentState => !currentState);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          currentTodoIds => currentTodoIds
+            .filter(currentTodoId => currentTodoId !== id),
+        );
+      });
+  };
+
+  const handleCompleteTodo = (todoId: number, completed: boolean) => {
+    setProcessingTodoIds(currentTodoIds => [...currentTodoIds, todoId]);
+
+    return toggleCompleteTodo(todoId, completed)
+      .then(() => {
+        setTodos(currentTodos => currentTodos.map(todo => (todo.id === todoId
+          ? { ...todo, completed }
+          : todo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to complete a todo');
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          currentTodoIds => currentTodoIds
+            .filter(currentTodoId => currentTodoId !== todoId),
+        );
+      });
+  };
+
+  const handleChangeTodoTitle = (todoId: number, newTitle: string) => {
+    setProcessingTodoIds(currentTodoIds => [...currentTodoIds, todoId]);
+
+    return changeTodoTitle(todoId, newTitle)
+      .then(() => {
+        setTodos(todos.map(todo => (todo.id === todoId
+          ? { ...todo, title: newTitle }
+          : todo
+        )));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+
+        throw new Error('Error: Unable to update a todo');
+      })
+      .finally(() => {
+        setProcessingTodoIds(
+          currentTodoIds => currentTodoIds
+            .filter(currentTodoId => currentTodoId !== todoId),
+        );
+      });
+  };
+
+  const handleToggleAll = () => {
+    const isAllTodosCompleted = todos.every(todo => todo.completed);
+
+    if (isAllTodosCompleted) {
+      todos.forEach(todo => {
+        handleCompleteTodo(todo.id, false);
+      });
+    } else {
+      todos.forEach(todo => {
+        if (todo.completed === false) {
```

Comment:
```suggestion
        if (!todo.completed) {
```

also can be simplify, check please

Reply:
also here